docs/architecture/adr-053-state-sync-prototype.md

```diff
@@ -192,14 +191,96 @@ Taking consistent snapshots of IAVL trees is greatly simplified by them being ve
 
 Snapshots must also be garbage collected after some configurable time, e.g. by keeping the latest `n` snapshots.
 
-## Open Questions
+## Experimental Prototype
 
-* Is it possible to reconstruct an identical IAVL tree given separate subtrees in an appropriate order, or is more data needed about the branch structure?
+An experimental but functional state sync prototype is available in the `erik/statesync-prototype` branches of the Tendermint, IAVL, Cosmos SDK, and Gaia repositories. To fetch the necessary branches:
 
-* Should we punish nodes that provide invalid snapshots?
+```sh
+$ mkdir statesync
+$ cd statesync
+$ git clone git@github.com:tendermint/tendermint -b erik/statesync-prototype
+$ git clone git@github.com:tendermint/iavl -b erik/statesync-prototype
+$ git clone git@github.com:cosmos/cosmos-sdk -b erik/statesync-prototype
+$ git clone git@github.com:cosmos/gaia -b erik/statesync-prototype
+```
+
+To spin up three nodes of a four-node testnet:
+
+```sh
+$ cd gaia
+$ ./tools/start.sh
+```
+
+Wait for the first snapshot to be taken at height 3, then (in a separate terminal) start the fourth node with state sync enabled:
+
+```sh
+$ ./tools/sync.sh
+```
+
+To stop the testnet, run:
+
+```sh
+$ ./tools/stop.sh
+```
+
+## Open Questions
 
 * Is it OK for state-synced nodes to not have historical blocks nor historical IAVL versions?
 
+* Do we need incremental chunk verification for first version?
+
+* Should we call these snapshots? IAVL already has a different concept called snapshots.
+
+* Should the snapshot ABCI interface be a separate optional ABCI service, or mandatory?
+
+* How can we make sure `ListSnapshots` data is valid? An adversary can provide a large number of invalid snapshots and make the node stuck in discovery mode. Perhaps collect snapshot list from many peers and pick ones that are available on a majority.
+
+* Should we punish nodes that provide invalid snapshots? How?
+
+* Should we store snapshot and chunk metadata in a database? Can we use the database for chunks?
+
+* How large chunks can we send without getting memory issues? Can we stream the chunk contents?
+
+* Should a snapshot at height H be taken before or after the block at H is processed? E.g. RPC `/commit` returns app_hash after _previous_ height, i.e. _before_  current height.
+
+* Do we need to support all versions of blockchain reactor (i.e. fast sync)?
```

Comment:
what work needs to be done to support each reactor version?

Reply:
Depends on how we decide to go forward with implementation. If we do a staged startup with dynamic reactor startup, then no additional work per version. Otherwise, we will need to implement switching from state sync to fast sync by manipulating reactor internals, which I think should be about 1-3 days for all versions - this is also somewhat more brittle, since we must support several combinations of reactor state transitions. Personally I think we should support all versions regardless, or possibly remove some reactor versions.

Reply:
> Personally I think we should support all versions regardless, or possibly remove some reactor versions.

agree. we should probably remove v1 since it's an intermediate version afaik